File: src/engine/extractor.ts
import {spawn} from 'node:child_process'
import fs from 'node:fs/promises'
import os from 'node:os'
import path from 'node:path'
import {formatBytes} from '../core/units.js'

export type VideoMeta = {
  title: string
  uploader?: string
  duration?: number
  webpage_url?: string
  extractor_key?: string
  formats?: RawStream[]
}

type RawStream = {
  format_id: string
  ext?: string
  vcodec?: string
  acodec?: string
  height?: number
  width?: number
  fps?: number
  dynamic_range?: string
  vbr?: number
  abr?: number
  tbr?: number
  filesize?: number
  filesize_approx?: number
}

export type ProbeResult = {
  meta: VideoMeta
  cachedJsonPath: string
}

export type Portion = {
  label: string
  kind: 'video' | 'audio'
  ytdlpArgs: string[]
}

export type ExtractPortionsOptions = {
  embedMetadata?: boolean
  videoContainer?: 'mp4' | 'mkv' | 'webm'
  audioFormat?: 'mp3' | 'm4a' | 'opus' | 'flac' | 'wav'
}

export async function probeVideo(
  ytdlpBin: string,
  url: string,
  signal?: AbortSignal,
): Promise<ProbeResult> {
  const stdout = await new Promise<string>((resolve, reject) => {
    const child = spawn(ytdlpBin, ['-J', '--no-playlist', '--no-warnings', url], {signal})
    let out = ''
    let errOut = ''
    child.stdout.on('data', chunk => (out += chunk))
    child.stderr.on('data', chunk => (errOut += chunk))
    child.on('error', reject)
    child.on('close', code => {
      if (code !== 0) {
        reject(new Error(cleanErrorOutput(errOut) || `yt-dlp exited with code ${code}`))
      } else {
        resolve(out)
      }
    })
  })

  let meta: VideoMeta
  try {
    meta = JSON.parse(stdout) as VideoMeta
  } catch {
    throw new Error('Could not parse video info from yt-dlp.')
  }

  // Dump full -J JSON to a temp file so bakeVideo can invoke --load-info-json.
  // Re-extracting from YouTube/X on download start costs 2-4 seconds of redundant network latency.
  const cachedJsonPath = path.join(os.tmpdir(), `anpan-meta-${process.pid}-${Date.now()}.json`)
  await fs.writeFile(cachedJsonPath, stdout)
  return {meta, cachedJsonPath}
}

export function extractPortions(meta: VideoMeta, opts?: ExtractPortionsOptions): Portion[] {
  const streams = meta.formats ?? []
  const portions: Portion[] = []
  const container = opts?.videoContainer ?? 'mp4'
  const audioFmt = opts?.audioFormat ?? 'mp3'

  const audioStreams = streams.filter(
    s => s.acodec && s.acodec !== 'none' && (!s.vcodec || s.vcodec === 'none'),
  )
  const bestAudio = [...audioStreams].sort((a, b) => (b.abr ?? b.tbr ?? 0) - (a.abr ?? a.tbr ?? 0))[0]
  let audioSize = bestAudio?.filesize ?? bestAudio?.filesize_approx
  if (!audioSize && (bestAudio?.abr || bestAudio?.tbr) && meta.duration) {
    audioSize = Math.round((((bestAudio.abr ?? bestAudio.tbr ?? 128) * 1000) / 8) * meta.duration)
  }

  const videoStreams = streams.filter(s => s.vcodec && s.vcodec !== 'none' && s.height)
  // Retrieve ALL available heights in descending order (no truncation to 8 tiers)
  const heights = [...new Set(videoStreams.map(s => s.height as number))].sort((a, b) => b - a)

  for (const height of heights) {
    const candidates = videoStreams.filter(s => s.height === height)
    const best = [...candidates].sort((a, b) => scoreStream(b, container) - scoreStream(a, container))[0]!
    const isMuxed = best.acodec !== undefined && best.acodec !== 'none'

    // Calculate video bytes from filesize, filesize_approx, or tbr/vbr
    let videoBytes = best.filesize ?? best.filesize_approx
    if (!videoBytes && (best.tbr || best.vbr) && meta.duration) {
      const rate = best.tbr ?? (best.vbr ?? 0) + (best.abr ?? 0)
      videoBytes = Math.round(((rate * 1000) / 8) * meta.duration)
    }

    const estimatedSize = videoBytes ? videoBytes + (isMuxed ? 0 : audioSize ?? 0) : 0
    const sizeTag = estimatedSize > 0 ? ` · ~${formatBytes(estimatedSize)}` : ''

    // Framerate & HDR tags
    const fpsTag = best.fps && best.fps >= 50 ? `${Math.round(best.fps)}` : ''
    const hdrTag = best.dynamic_range && best.dynamic_range.toLowerCase().includes('hdr') ? ' HDR' : ''

    portions.push({
      kind: 'video',
      label: `${height}p${fpsTag}${hdrTag} · ${container}${sizeTag}`,
      ytdlpArgs: [
        '-f',
        `bv*[height=${height}]+ba/b[height=${height}]/bv*[height<=${height}]+ba/b`,
        '--merge-output-format',
        container,
      ],
    })
  }

  if (portions.length === 0) {
    portions.push({
      kind: 'video',
      label: `best available · ${container}`,
      ytdlpArgs: ['-f', 'bv*+ba/b', '--merge-output-format', container],
    })
  }

  // Audio streams
  const addedAudioFormats = new Set<string>()

  // 1. Check for native Opus stream (YouTube / YTM native audio)
  const nativeOpus = audioStreams.find(
    s => s.acodec?.toLowerCase().includes('opus') || s.ext === 'webm',
  )
  if (nativeOpus) {
    let opusBytes = nativeOpus.filesize ?? nativeOpus.filesize_approx
    if (!opusBytes && (nativeOpus.abr || nativeOpus.tbr) && meta.duration) {
      opusBytes = Math.round((((nativeOpus.abr ?? nativeOpus.tbr ?? 128) * 1000) / 8) * meta.duration)
    }
    const opusSizeTag = opusBytes ? ` · ~${formatBytes(opusBytes)}` : ''
    const opusAbrTag = nativeOpus.abr ? ` · ~${Math.round(nativeOpus.abr)}kbps` : ''
    const opusArgs = ['-f', 'ba[acodec^=opus]/ba', '-x', '--audio-format', 'opus']
    if (opts?.embedMetadata !== false) opusArgs.push('--embed-thumbnail', '--add-metadata')
    portions.push({
      kind: 'audio',
      label: `audio only · opus (original)${opusAbrTag}${opusSizeTag}`,
      ytdlpArgs: opusArgs,
    })
    addedAudioFormats.add('opus')
  }

  // 2. Check for native M4A / AAC stream
  const nativeM4a = audioStreams.find(
    s => s.ext === 'm4a' || s.acodec?.toLowerCase().includes('mp4a'),
  )
  if (nativeM4a) {
    let m4aBytes = nativeM4a.filesize ?? nativeM4a.filesize_approx
    if (!m4aBytes && (nativeM4a.abr || nativeM4a.tbr) && meta.duration) {
      m4aBytes = Math.round((((nativeM4a.abr ?? nativeM4a.tbr ?? 128) * 1000) / 8) * meta.duration)
    }
    const m4aSizeTag = m4aBytes ? ` · ~${formatBytes(m4aBytes)}` : ''
    const m4aAbrTag = nativeM4a.abr ? ` · ~${Math.round(nativeM4a.abr)}kbps` : ''
    const m4aArgs = ['-f', 'ba[ext=m4a]/ba', '-x', '--audio-format', 'm4a']
    if (opts?.embedMetadata !== false) m4aArgs.push('--embed-thumbnail', '--add-metadata')
    portions.push({
      kind: 'audio',
      label: `audio only · m4a (aac)${m4aAbrTag}${m4aSizeTag}`,
      ytdlpArgs: m4aArgs,
    })
    addedAudioFormats.add('m4a')
  }

  // 3. User configured audio format (if not already added as opus/m4a)
  if (!addedAudioFormats.has(audioFmt)) {
    const audioSizeTag = audioSize ? ` · ~${formatBytes(audioSize)}` : ''
    const audioBitrateTag = bestAudio?.abr ? ` · ${Math.round(bestAudio.abr)}kbps` : ''
    const audioQuality = audioFmt === 'flac' || audioFmt === 'wav' ? '0' : '0'
    const audioArgs = ['-f', 'ba/b', '-x', '--audio-format', audioFmt, '--audio-quality', audioQuality]
    if (opts?.embedMetadata !== false) audioArgs.push('--embed-thumbnail', '--add-metadata')
    portions.push({
      kind: 'audio',
      label: `audio only · ${audioFmt}${audioBitrateTag}${audioSizeTag}`,
      ytdlpArgs: audioArgs,
    })
    addedAudioFormats.add(audioFmt)
  }

  // 4. Universal MP3 fallback if not yet added
  if (!addedAudioFormats.has('mp3')) {
    const mp3SizeTag = audioSize ? ` · ~${formatBytes(audioSize)}` : ''
    const mp3Args = ['-f', 'ba/b', '-x', '--audio-format', 'mp3', '--audio-quality', '0']
    if (opts?.embedMetadata !== false) mp3Args.push('--embed-thumbnail', '--add-metadata')
    portions.push({
      kind: 'audio',
      label: `audio only · mp3 · 320kbps${mp3SizeTag}`,
      ytdlpArgs: mp3Args,
    })
    addedAudioFormats.add('mp3')
  }

  return portions
}

export type PlaylistMeta = {
  title: string
  uploader?: string
  trackCount: number
  webpage_url: string
}

// Light probe for playlist info using --flat-playlist: returns in ~1s without pulling full stream lists
export async function probePlaylist(
  ytdlpBin: string,
  url: string,
  signal?: AbortSignal,
): Promise<PlaylistMeta | null> {
  try {
    const stdout = await new Promise<string>((resolve, reject) => {
      const child = spawn(
        ytdlpBin,
        ['--flat-playlist', '-J', '--no-warnings', url],
        {signal},
      )
      let out = ''
      let errOut = ''
      child.stdout.on('data', chunk => (out += chunk))
      child.stderr.on('data', chunk => (errOut += chunk))
      child.on('error', reject)
      child.on('close', code => {
        if (code !== 0) reject(new Error(cleanErrorOutput(errOut) || `yt-dlp exited with code ${code}`))
        else resolve(out)
      })
    })

    const data = JSON.parse(stdout) as {
      _type?: string
      title?: string
      uploader?: string
      playlist_count?: number
      entries?: unknown[]
      webpage_url?: string
    }

    if (data._type === 'playlist' || Array.isArray(data.entries)) {
      const trackCount = data.playlist_count ?? data.entries?.length ?? 0
      return {
        title: data.title || 'Playlist',
        uploader: data.uploader,
        trackCount,
        webpage_url: data.webpage_url || url,
      }
    }
    return null
  } catch {
    return null
  }
}

export function extractPlaylistPortions(opts?: ExtractPortionsOptions): Portion[] {
  const container = opts?.videoContainer ?? 'mp4'
  const audioFmt = opts?.audioFormat ?? 'mp3'

  const audioQuality = audioFmt === 'flac' || audioFmt === 'wav' ? '0' : '0'
  const audioArgs = ['-f', 'ba/b', '-x', '--audio-format', audioFmt, '--audio-quality', audioQuality]
  if (opts?.embedMetadata !== false) {
    audioArgs.push('--embed-thumbnail', '--add-metadata')
  }
  const portions: Portion[] = [
    {
      kind: 'audio',
      label: `all tracks · ${audioFmt} (audio only)`,
      ytdlpArgs: audioArgs,
    },
  ]

  if (audioFmt !== 'opus') {
    const opusArgs = ['-f', 'ba[acodec^=opus]/ba', '-x', '--audio-format', 'opus']
    if (opts?.embedMetadata !== false) opusArgs.push('--embed-thumbnail', '--add-metadata')
    portions.push({
      kind: 'audio',
      label: 'all tracks · opus (original audio)',
      ytdlpArgs: opusArgs,
    })
  }

  if (audioFmt !== 'mp3') {
    const mp3Args = ['-f', 'ba/b', '-x', '--audio-format', 'mp3', '--audio-quality', '0']
    if (opts?.embedMetadata !== false) mp3Args.push('--embed-thumbnail', '--add-metadata')
    portions.push({
      kind: 'audio',
      label: 'all tracks · mp3 (audio only)',
      ytdlpArgs: mp3Args,
    })
  }

  portions.push({
    kind: 'video',
    label: `all videos · ${container} (best quality)`,
    ytdlpArgs: ['-f', 'bv*+ba/b', '--merge-output-format', container],
  })

  return portions
}

function scoreStream(s: RawStream, preferredContainer = 'mp4'): number {
  let score = s.tbr ?? 0
  if (s.ext === preferredContainer) score += 10_000
  if (s.vcodec?.startsWith('avc')) score += 5_000
  return score
}

function cleanErrorOutput(stderr: string): string {
  const lines = stderr
    .split('\n')
    .map(l => l.trim())
    .filter(l => l.startsWith('ERROR:'))
  const last = lines.at(-1)
  return last ? last.replace(/^ERROR:\s*(\[[^\]]+\]\s*)?/, '') : ''
}

File: internal/engine/extractor.go
package engine

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"os/exec"
	"regexp"
	"sort"
	"strings"

	"github.com/KabosuNeko/anpan/internal/units"
)

type RawStream struct {
	FormatID       string   `json:"format_id"`
	Ext            string   `json:"ext,omitempty"`
	Vcodec         string   `json:"vcodec,omitempty"`
	Acodec         string   `json:"acodec,omitempty"`
	Height         *int     `json:"height,omitempty"`
	Width          *int     `json:"width,omitempty"`
	FPS            *float64 `json:"fps,omitempty"`
	DynamicRange   string   `json:"dynamic_range,omitempty"`
	VBR            *float64 `json:"vbr,omitempty"`
	ABR            *float64 `json:"abr,omitempty"`
	TBR            *float64 `json:"tbr,omitempty"`
	Filesize       *float64 `json:"filesize,omitempty"`
	FilesizeApprox *float64 `json:"filesize_approx,omitempty"`
}

type VideoMeta struct {
	Title        string      `json:"title"`
	Uploader     string      `json:"uploader,omitempty"`
	Duration     *float64    `json:"duration,omitempty"`
	WebpageURL   string      `json:"webpage_url,omitempty"`
	ExtractorKey string      `json:"extractor_key,omitempty"`
	Formats      []RawStream `json:"formats,omitempty"`
}

type PortionKind string

const (
	PortionKindVideo PortionKind = "video"
	PortionKindAudio PortionKind = "audio"
)

type Portion struct {
	Label     string      `json:"label"`
	Kind      PortionKind `json:"kind"`
	YtdlpArgs []string    `json:"ytdlpArgs"`
}

type ExtractPortionsOptions struct {
	EmbedMetadata  *bool
	VideoContainer string // "mp4" | "mkv" | "webm"
	VideoCodec     string // "auto" | "av1" | "vp9" | "avc"
	AudioFormat    string // "mp3" | "m4a" | "opus" | "flac" | "wav"
}

type ProbeResult struct {
	Meta           VideoMeta
	CachedJSONPath string
}

func CleanErrorOutput(stderr string) string {
	scanner := bufio.NewScanner(strings.NewReader(stderr))
	var errorLines []string
	re := regexp.MustCompile(`^ERROR:\s*(?:\[[^\]]+\]\s*)?`)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if strings.HasPrefix(line, "ERROR:") {
			errorLines = append(errorLines, re.ReplaceAllString(line, ""))
		}
	}
	if len(errorLines) > 0 {
		return errorLines[len(errorLines)-1]
	}
	return ""
}

func ProbeVideo(ctx context.Context, ytdlpBin string, rawURL string) (*ProbeResult, error) {
	cmd := exec.CommandContext(ctx, ytdlpBin, "-J", "--no-playlist", "--no-warnings", rawURL)
	out, err := cmd.Output()
	if err != nil {
		stderr := ""
		if exitErr, ok := err.(*exec.ExitError); ok {
			stderr = string(exitErr.Stderr)
		}
		clean := CleanErrorOutput(stderr)
		if clean != "" {
			return nil, fmt.Errorf("%s", clean)
		}
		return nil, fmt.Errorf("could not probe video: %w", err)
	}

	var meta VideoMeta
	if err := json.Unmarshal(out, &meta); err != nil {
		return nil, fmt.Errorf("could not parse media info from yt-dlp: %w", err)
	}

	tmpFile, err := os.CreateTemp("", "anpan-meta-*.json")
	if err != nil {
		return nil, err
	}
	if _, err := tmpFile.Write(out); err != nil {
		_ = tmpFile.Close()
		return nil, err
	}
	_ = tmpFile.Close()

	return &ProbeResult{
		Meta:           meta,
		CachedJSONPath: tmpFile.Name(),
	}, nil
}

func scoreStream(s RawStream, preferredContainer string, preferredCodec string) float64 {
	score := float64(0)
	if s.TBR != nil {
		score = *s.TBR
	}
	if s.Ext == preferredContainer {
		score += 1000
	}
	switch preferredCodec {
	case "av1":
		if strings.HasPrefix(s.Vcodec, "av01") {
			score += 50000
		}
	case "vp9":
		if strings.HasPrefix(s.Vcodec, "vp09") || strings.HasPrefix(s.Vcodec, "vp9") {
			score += 50000
		}
	case "avc":
		if strings.HasPrefix(s.Vcodec, "avc") || strings.HasPrefix(s.Vcodec, "h264") {
			score += 50000
		}
	default:
		// "auto": prioritize quality-per-bitrate: AV1 > VP9 > AVC
		if strings.HasPrefix(s.Vcodec, "av01") {
			score += 3000
		} else if strings.HasPrefix(s.Vcodec, "vp09") || strings.HasPrefix(s.Vcodec, "vp9") {
			score += 2000
		} else if strings.HasPrefix(s.Vcodec, "avc") || strings.HasPrefix(s.Vcodec, "h264") {
			score += 1000
		}
	}
	return score
}

func ExtractPortions(meta VideoMeta, opts *ExtractPortionsOptions) []Portion {
	streams := meta.Formats
	var portions []Portion

	container := "mp4"
	audioFmt := "mp3"
	codec := "auto"
	embedMetadata := true
	if opts != nil {
		if opts.VideoContainer != "" {
			container = opts.VideoContainer
		}
		if opts.AudioFormat != "" {
			audioFmt = opts.AudioFormat
		}
		if opts.VideoCodec != "" {
			codec = opts.VideoCodec
		}
		if opts.EmbedMetadata != nil {
			embedMetadata = *opts.EmbedMetadata
		}
	}

	var audioStreams []RawStream
	for _, s := range streams {
		if s.Acodec != "" && s.Acodec != "none" && (s.Vcodec == "" || s.Vcodec == "none") {
			audioStreams = append(audioStreams, s)
		}
	}

	var bestAudio *RawStream
	for i := range audioStreams {
		s := &audioStreams[i]
		if bestAudio == nil {
			bestAudio = s
			continue
		}
		sRate := float64(0)
		if s.ABR != nil {
			sRate = *s.ABR
		} else if s.TBR != nil {
			sRate = *s.TBR
		}
		bestRate := float64(0)
		if bestAudio.ABR != nil {
			bestRate = *bestAudio.ABR
		} else if bestAudio.TBR != nil {
			bestRate = *bestAudio.TBR
		}
		if sRate > bestRate {
			bestAudio = s
		}
	}

	var audioSize *float64
	if bestAudio != nil {
		if bestAudio.Filesize != nil {
			audioSize = bestAudio.Filesize
		} else if bestAudio.FilesizeApprox != nil {
			audioSize = bestAudio.FilesizeApprox
		} else if (bestAudio.ABR != nil || bestAudio.TBR != nil) && meta.Duration != nil {
			rate := float64(128)
			if bestAudio.ABR != nil {
				rate = *bestAudio.ABR
			} else if bestAudio.TBR != nil {
				rate = *bestAudio.TBR
			}
			calc := math.Round(((rate * 1000) / 8) * (*meta.Duration))
			audioSize = &calc
		}
	}

	hasVideo := false
	var videoStreams []RawStream
	heightMap := make(map[int]bool)
	for _, s := range streams {
		if s.Vcodec != "" && s.Vcodec != "none" {
			hasVideo = true
			if s.Height != nil && *s.Height > 0 {
				videoStreams = append(videoStreams, s)
				heightMap[*s.Height] = true
			}
		}
	}

	var heights []int
	for h := range heightMap {
		heights = append(heights, h)
	}
	sort.Slice(heights, func(i, j int) bool {
		return heights[i] > heights[j]
	})

	for _, height := range heights {
		var candidates []RawStream
		for _, s := range videoStreams {
			if s.Height != nil && *s.Height == height {
				candidates = append(candidates, s)
			}
		}
		if len(candidates) == 0 {
			continue
		}
		sort.Slice(candidates, func(i, j int) bool {
			return scoreStream(candidates[i], container, codec) > scoreStream(candidates[j], container, codec)
		})
		best := candidates[0]
		isMuxed := best.Acodec != "" && best.Acodec != "none"

		var videoBytes *float64
		if best.Filesize != nil {
			videoBytes = best.Filesize
		} else if best.FilesizeApprox != nil {
			videoBytes = best.FilesizeApprox
		} else if (best.TBR != nil || best.VBR != nil) && meta.Duration != nil {
			rate := float64(0)
			if best.TBR != nil {
				rate = *best.TBR
			} else {
				if best.VBR != nil {
					rate += *best.VBR
				}
				if best.ABR != nil {
					rate += *best.ABR
				}
			}
			calc := math.Round(((rate * 1000) / 8) * (*meta.Duration))
			videoBytes = &calc
		}

		estimatedSize := float64(0)
		if videoBytes != nil {
			estimatedSize = *videoBytes
			if !isMuxed && audioSize != nil {
				estimatedSize += *audioSize
			}
		}

		sizeTag := ""
		if estimatedSize > 0 {
			sizeTag = fmt.Sprintf(" · ~%s", units.FormatBytes(estimatedSize))
		}
		fpsTag := ""
		if best.FPS != nil && *best.FPS >= 50 {
			fpsTag = fmt.Sprintf("%.0f", math.Round(*best.FPS))
		}
		hdrTag := ""
		if strings.Contains(strings.ToLower(best.DynamicRange), "hdr") {
			hdrTag = " HDR"
		}

		var ytdlpSelector string
		switch codec {
		case "av1":
			ytdlpSelector = fmt.Sprintf("bv*[height=%d][vcodec^=av01]+ba/bv*[height=%d]+ba/b[height=%d]/bv*[height<=%d]+ba/b", height, height, height, height)
		case "vp9":
			ytdlpSelector = fmt.Sprintf("bv*[height=%d][vcodec^=vp09]+ba/bv*[height=%d]+ba/b[height=%d]/bv*[height<=%d]+ba/b", height, height, height, height)
		case "avc":
			ytdlpSelector = fmt.Sprintf("bv*[height=%d][vcodec^=avc1]+ba/bv*[height=%d]+ba/b[height=%d]/bv*[height<=%d]+ba/b", height, height, height, height)
		default:
			ytdlpSelector = fmt.Sprintf("bv*[height=%d]+ba/b[height=%d]/bv*[height<=%d]+ba/b", height, height, height)
		}

		portions = append(portions, Portion{
			Kind:  PortionKindVideo,
			Label: fmt.Sprintf("%dp%s%s · %s%s", height, fpsTag, hdrTag, container, sizeTag),
			YtdlpArgs: []string{
				"-f",
				ytdlpSelector,
				"--merge-output-format",
				container,
			},
		})
	}

	if hasVideo && len(portions) == 0 {
		portions = append(portions, Portion{
			Kind:      PortionKindVideo,
			Label:     fmt.Sprintf("best available · %s", container),
			YtdlpArgs: []string{"-f", "bv*+ba/b", "--merge-output-format", container},
		})
	}

	addedAudioFormats := make(map[string]bool)
	nativeAudioMap := make(map[string]RawStream)

	for _, s := range audioStreams {
		codec := strings.ToLower(s.Acodec)
		ext := strings.ToLower(s.Ext)
		key := ""
		if strings.Contains(codec, "flac") || ext == "flac" {
			key = "flac"
		} else if strings.Contains(codec, "opus") || ext == "webm" {
			key = "opus"
		} else if strings.Contains(codec, "mp4a") || strings.Contains(codec, "aac") || ext == "m4a" {
			key = "m4a"
		} else if strings.Contains(codec, "mp3") || ext == "mp3" {
			key = "mp3"
		} else if strings.Contains(codec, "vorbis") || ext == "ogg" {
			key = "ogg"
		} else if ext != "" && ext != "none" {
			key = ext
		}

		if key != "" {
			existing, exists := nativeAudioMap[key]
			sRate := float64(0)
			if s.ABR != nil {
				sRate = *s.ABR
			} else if s.TBR != nil {
				sRate = *s.TBR
			}
			exRate := float64(0)
			if existing.ABR != nil {
				exRate = *existing.ABR
			} else if existing.TBR != nil {
				exRate = *existing.TBR
			}
			if !exists || sRate > exRate {
				nativeAudioMap[key] = s
			}
		}
	}

	priorityKeys := []string{"opus", "flac", "m4a", "mp3", "ogg"}
	var orderedKeys []string
	for _, pk := range priorityKeys {
		if _, ok := nativeAudioMap[pk]; ok {
			orderedKeys = append(orderedKeys, pk)
		}
	}
	for k := range nativeAudioMap {
		found := false
		for _, pk := range priorityKeys {
			if pk == k {
				found = true
				break
			}
		}
		if !found {
			orderedKeys = append(orderedKeys, k)
		}
	}

	for _, key := range orderedKeys {
		stream := nativeAudioMap[key]
		var bytes *float64
		if stream.Filesize != nil {
			bytes = stream.Filesize
		} else if stream.FilesizeApprox != nil {
			bytes = stream.FilesizeApprox
		} else if (stream.ABR != nil || stream.TBR != nil) && meta.Duration != nil {
			rate := float64(128)
			if stream.ABR != nil {
				rate = *stream.ABR
			} else if stream.TBR != nil {
				rate = *stream.TBR
			}
			calc := math.Round(((rate * 1000) / 8) * (*meta.Duration))
			bytes = &calc
		}

		sizeTag := ""
		if bytes != nil {
			sizeTag = fmt.Sprintf(" · ~%s", units.FormatBytes(*bytes))
		}
		abrTag := ""
		if stream.ABR != nil {
			abrTag = fmt.Sprintf(" · ~%.0fkbps", math.Round(*stream.ABR))
		}
		isLossless := key == "flac" || key == "wav"
		tagSuffix := " (original)"
		if isLossless {
			tagSuffix = " (lossless)"
		} else if key == "m4a" {
			tagSuffix = " (aac)"
		}

		ytdlpSelector := fmt.Sprintf("ba[ext=%s]/ba", stream.Ext)
		if key == "opus" {
			ytdlpSelector = "ba[acodec^=opus]/ba"
		} else if key == "m4a" {
			ytdlpSelector = "ba[ext=m4a]/ba"
		} else if stream.FormatID != "" {
			ytdlpSelector = fmt.Sprintf("%s/ba", stream.FormatID)
		}

		audioArgs := []string{"-f", ytdlpSelector, "-x", "--audio-format", key}
		if embedMetadata {
			if key != "wav" {
				audioArgs = append(audioArgs, "--embed-thumbnail")
			}
			audioArgs = append(audioArgs, "--add-metadata")
		}

		portions = append(portions, Portion{
			Kind:      PortionKindAudio,
			Label:     fmt.Sprintf("audio only · %s%s%s%s", key, tagSuffix, abrTag, sizeTag),
			YtdlpArgs: audioArgs,
		})
		addedAudioFormats[key] = true
	}

	if !addedAudioFormats[audioFmt] {
		audioSizeTag := ""
		if audioSize != nil {
			audioSizeTag = fmt.Sprintf(" · ~%s", units.FormatBytes(*audioSize))
		}
		audioBitrateTag := ""
		if bestAudio != nil && bestAudio.ABR != nil {
			audioBitrateTag = fmt.Sprintf(" · %.0fkbps", math.Round(*bestAudio.ABR))
		}
		audioArgs := []string{"-f", "ba/b", "-x", "--audio-format", audioFmt, "--audio-quality", "0"}
		if embedMetadata {
			if audioFmt != "wav" {
				audioArgs = append(audioArgs, "--embed-thumbnail")
			}
			audioArgs = append(audioArgs, "--add-metadata")
		}
		portions = append(portions, Portion{
			Kind:      PortionKindAudio,
			Label:     fmt.Sprintf("audio only · %s%s%s", audioFmt, audioBitrateTag, audioSizeTag),
			YtdlpArgs: audioArgs,
		})
		addedAudioFormats[audioFmt] = true
	}

	if !addedAudioFormats["mp3"] {
		mp3SizeTag := ""
		if audioSize != nil {
			mp3SizeTag = fmt.Sprintf(" · ~%s", units.FormatBytes(*audioSize))
		}
		mp3Args := []string{"-f", "ba/b", "-x", "--audio-format", "mp3", "--audio-quality", "0"}
		if embedMetadata {
			mp3Args = append(mp3Args, "--embed-thumbnail", "--add-metadata")
		}
		portions = append(portions, Portion{
			Kind:      PortionKindAudio,
			Label:     fmt.Sprintf("audio only · mp3 · 320kbps%s", mp3SizeTag),
			YtdlpArgs: mp3Args,
		})
		addedAudioFormats["mp3"] = true
	}

	if len(portions) == 0 {
		portions = append(portions, Portion{
			Kind:      PortionKindVideo,
			Label:     fmt.Sprintf("best available · %s", container),
			YtdlpArgs: []string{"-f", "bv*+ba/b", "--merge-output-format", container},
		})
	}

	return portions
}

var albumPrefixRegex = regexp.MustCompile(`(?i)^(?:album|ep|single)\s*[-:–—]\s*`)

func CleanAlbumTitle(title string) string {
	if strings.TrimSpace(title) == "" {
		return "Playlist"
	}
	cleaned := albumPrefixRegex.ReplaceAllString(strings.TrimSpace(title), "")
	if strings.TrimSpace(cleaned) == "" {
		return title
	}
	return strings.TrimSpace(cleaned)
}

type PlaylistMeta struct {
	Title      string `json:"title"`
	Uploader   string `json:"uploader,omitempty"`
	TrackCount int    `json:"trackCount"`
	WebpageURL string `json:"webpage_url"`
}

func ProbePlaylist(ctx context.Context, ytdlpBin string, rawURL string) (*PlaylistMeta, error) {
	cmd := exec.CommandContext(ctx, ytdlpBin, "--flat-playlist", "-J", "--no-warnings", rawURL)
	out, err := cmd.Output()
	if err != nil {
		return nil, err
	}

	var data struct {
		Type          string `json:"_type"`
		Title         string `json:"title"`
		Uploader      string `json:"uploader"`
		PlaylistCount *int   `json:"playlist_count"`
		Entries       []any  `json:"entries"`
		WebpageURL    string `json:"webpage_url"`
	}

	if err := json.Unmarshal(out, &data); err != nil {
		return nil, err
	}

	if data.Type == "playlist" || len(data.Entries) > 0 {
		count := len(data.Entries)
		if data.PlaylistCount != nil {
			count = *data.PlaylistCount
		}
		webURL := data.WebpageURL
		if webURL == "" {
			webURL = rawURL
		}
		return &PlaylistMeta{
			Title:      CleanAlbumTitle(data.Title),
			Uploader:   data.Uploader,
			TrackCount: count,
			WebpageURL: webURL,
		}, nil
	}

	return nil, fmt.Errorf("not a playlist")
}

func ExtractPlaylistPortions(opts *ExtractPortionsOptions) []Portion {
	container := "mp4"
	audioFmt := "mp3"
	codec := "auto"
	embedMetadata := true
	if opts != nil {
		if opts.VideoContainer != "" {
			container = opts.VideoContainer
		}
		if opts.AudioFormat != "" {
			audioFmt = opts.AudioFormat
		}
		if opts.VideoCodec != "" {
			codec = opts.VideoCodec
		}
		if opts.EmbedMetadata != nil {
			embedMetadata = *opts.EmbedMetadata
		}
	}

	audioArgs := []string{"-f", "ba/b", "-x", "--audio-format", audioFmt, "--audio-quality", "0"}
	if embedMetadata {
		audioArgs = append(audioArgs, "--embed-thumbnail", "--add-metadata")
	}

	portions := []Portion{
		{
			Kind:      PortionKindAudio,
			Label:     fmt.Sprintf("all tracks · %s (audio only)", audioFmt),
			YtdlpArgs: audioArgs,
		},
	}

	if audioFmt != "opus" {
		opusArgs := []string{"-f", "ba[acodec^=opus]/ba", "-x", "--audio-format", "opus"}
		if embedMetadata {
			opusArgs = append(opusArgs, "--embed-thumbnail", "--add-metadata")
		}
		portions = append(portions, Portion{
			Kind:      PortionKindAudio,
			Label:     "all tracks · opus (original audio)",
			YtdlpArgs: opusArgs,
		})
	}

	if audioFmt != "mp3" {
		mp3Args := []string{"-f", "ba/b", "-x", "--audio-format", "mp3", "--audio-quality", "0"}
		if embedMetadata {
			mp3Args = append(mp3Args, "--embed-thumbnail", "--add-metadata")
		}
		portions = append(portions, Portion{
			Kind:      PortionKindAudio,
			Label:     "all tracks · mp3 (audio only)",
			YtdlpArgs: mp3Args,
		})
	}

	var videoSelector string
	switch codec {
	case "av1":
		videoSelector = "bv*[vcodec^=av01]+ba/bv*+ba/b"
	case "vp9":
		videoSelector = "bv*[vcodec^=vp09]+ba/bv*+ba/b"
	case "avc":
		videoSelector = "bv*[vcodec^=avc1]+ba/bv*+ba/b"
	default:
		videoSelector = "bv*+ba/b"
	}

	portions = append(portions, Portion{
		Kind:      PortionKindVideo,
		Label:     fmt.Sprintf("all videos · %s (best quality)", container),
		YtdlpArgs: []string{"-f", videoSelector, "--merge-output-format", container},
	})

	return portions
}
